File: python/31_NextPermutation.py
from unittest import TestCase
# https://leetcode.com/problems/next-permutation/


class NextPermutation(TestCase):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        lower = next((i for i in range(len(nums)-2, -1, -1) if nums[i] < nums[i+1]), -1)
        if lower >= 0:
            k = next((i for i in range(len(nums)-1, lower, -1) if nums[i] > nums[lower]))
            nums[lower], nums[k] = nums[k], nums[lower]

        i, j = lower+1, len(nums)-1
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1

    def test1(self):
        nums = [1, 2, 3]
        self.nextPermutation(nums)
        self.assertEqual(nums, [1, 3, 2])

    def test2(self):
        nums = [3, 2, 1]
        self.nextPermutation(nums)
        self.assertEqual(nums, [1, 2, 3])

    def test3(self):
        nums = [1, 1, 5]
        self.nextPermutation(nums)
        self.assertEqual(nums, [1, 5, 1])

File: kotlin/src/main/kotlin/31.NextPermutation.kt
import org.junit.Assert.assertEquals
import org.junit.Test

//https://leetcode.com/problems/next-permutation/

class NextPermutation{
  fun nextPermutation(nums: IntArray): Unit {
    val lower = (nums.size-2 downTo 0).find{ nums[it] < nums[it+1] }

    if(lower != null){
      val k = (nums.size-1 downTo lower+1).find{ nums[it] > nums[lower] }!!
      swap(nums, lower, k)
    }
    var l=(lower ?: -1)+1
    var r=nums.size-1
    while(l < r)
      swap(nums, l++, r--)
  }

  private fun swap(nums: IntArray, i: Int, j: Int){
    val tmp = nums[i]
    nums[i] = nums[j]
    nums[j] = tmp
  }

  @Test
  fun test1(){
    val nums = intArrayOf(1,2,3)
    nextPermutation(nums)
    assertEquals(listOf(1,3,2), nums.toList())
  }

  @Test
  fun test2(){
    val nums = intArrayOf(3,2,1)
    nextPermutation(nums)
    assertEquals(listOf(1,2,3), nums.toList())
  }

  @Test
  fun test3(){
    val nums = intArrayOf(1,1,5)
    nextPermutation(nums)
    assertEquals(listOf(1,5,1), nums.toList())
  }

}

